serde_derive/src/ser.rs

```diff
@@ -505,7 +513,13 @@ fn serialize_externally_tagged_variant(
     cattrs: &attr::Container,
 ) -> Fragment {
     let type_name = cattrs.name().serialize_name();
-    let variant_name = variant.attrs.name().serialize_name();
+
+    let variant_name = match variant.attrs.name().serialize_name() {
+        VariantName::String(name) => name,
+        // An externally tagged variant with a non-string name will fail to
+        // check, so this branch _should_ be unreachable.
```

Comment:
If you are confident about this, it should come across that way in the comment. If you are not confident about this, then this code would need to be changed.